pkg/scheduler/framework/plugins/podtopologyspread/scoring.go

```diff
@@ -0,0 +1,255 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podtopologyspread
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync/atomic"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/predicates"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+)
+
+const postFilterStateKey = "PostFilter" + Name
+
+// postFilterState computed at PostFilter and used at Score.
+type postFilterState struct {
+	constraints []topologySpreadConstraint
+	// nodeNameSet is a string set holding all node names which have all constraints[*].topologyKey present.
+	nodeNameSet map[string]struct{}
+	// topologyPairToPodCounts is keyed with topologyPair, and valued with the number of matching pods.
+	topologyPairToPodCounts map[topologyPair]*int64
+}
+
+// Clone implements the mandatory Clone interface. We don't really copy the data since
+// there is no need for that.
+func (s *postFilterState) Clone() framework.StateData {
+	return s
+}
+
+// initialize iterates "filteredNodes" to filter out the nodes which don't have required topologyKey(s),
+// and initialize two maps:
+// 1) s.topologyPairToPodCounts: keyed with both eligible topology pair and node names.
+// 2) s.nodeNameSet: keyed with node name, and valued with a *int64 pointer for eligible node only.
+func (s *postFilterState) initialize(pod *v1.Pod, filteredNodes []*v1.Node) error {
+	constraints, err := filterTopologySpreadConstraints(pod.Spec.TopologySpreadConstraints, v1.ScheduleAnyway)
+	if err != nil {
+		return err
+	}
+	if constraints == nil {
+		return nil
+	}
+	s.constraints = constraints
+	for _, node := range filteredNodes {
+		if !nodeLabelsMatchSpreadConstraints(node.Labels, s.constraints) {
+			continue
+		}
+		for _, constraint := range s.constraints {
+			pair := topologyPair{key: constraint.topologyKey, value: node.Labels[constraint.topologyKey]}
+			if s.topologyPairToPodCounts[pair] == nil {
+				s.topologyPairToPodCounts[pair] = new(int64)
+			}
+		}
+		s.nodeNameSet[node.Name] = struct{}{}
+		// For those nodes which don't have all required topologyKeys present, it's intentional to leave
+		// their entries absent in nodeNameSet, so that we're able to score them to 0 afterwards.
+	}
+	return nil
+}
+
+// PostFilter builds and writes cycle state used by Score and NormalizeScore.
+func (pl *PodTopologySpread) PostFilter(
+	ctx context.Context,
+	cycleState *framework.CycleState,
+	pod *v1.Pod,
+	filteredNodes []*v1.Node,
+	_ framework.NodeToStatusMap,
+) *framework.Status {
+	allNodes, err := pl.sharedLister.NodeInfos().List()
+	if err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("error when getting all nodes: %v", err))
+	}
+
+	state := &postFilterState{
```

Comment:
nit: move this closer to where it is first used (line 102).

Reply:
done.